runtime/major_gc.c

```diff
@@ -112,6 +113,84 @@ static struct ephe_cycle_info_t {
  * the lock. */
 static caml_plat_mutex ephe_lock = CAML_PLAT_MUTEX_INITIALIZER;
 
+#define PREFETCH_BUFFER_SIZE  (1 << 8)
+#define PREFETCH_BUFFER_MIN   64 /* keep pb at least this full */
+#define PREFETCH_BUFFER_MASK  (PREFETCH_BUFFER_SIZE - 1)
+
+typedef struct prefetch_buffer {
+  uintnat enqueued;
+  uintnat dequeued;
+  uintnat waterline;
+  value   buffer[PREFETCH_BUFFER_SIZE];
+} prefetch_buffer_t;
+
+Caml_inline bool pb_full(const prefetch_buffer_t *pb)
+{
+  return pb->enqueued == (pb->dequeued + PREFETCH_BUFFER_SIZE);
+}
+
+Caml_inline uintnat pb_size(const prefetch_buffer_t *pb)
+{
+  return pb->enqueued - pb->dequeued;
+}
+
+Caml_inline bool pb_above_waterline(const prefetch_buffer_t *pb)
+{
+  return ((pb->enqueued - pb->dequeued) > pb->waterline);
+}
+
+Caml_inline void pb_drain(prefetch_buffer_t *pb)
```

Comment:
I read `pb_fill` and `pb_drain` as actions, just like `pb_push` and `pb_pop`. I wonder if `pb_fill_mode` and `pb_drain_mode` would be clearer -- we are not draining or filling the queue, we are changing its mode to drain or fill over time.

Reply:
Agree, this may be confusing!